--- a/init	2025-07-10 22:57:57.810764118 -0700
+++ b/init	2025-07-11 12:28:39.069368881 -0700
@@ -6,6 +6,7 @@
 
 [ -d /dev ] || mkdir -m 0755 /dev
 [ -d /root ] || mkdir -m 0700 /root
+[ -d /root_ ] || mkdir -m 0700 /root_
 [ -d /sys ] || mkdir /sys
 [ -d /proc ] || mkdir /proc
 [ -d /tmp ] || mkdir /tmp
@@ -64,7 +65,7 @@
 export break=
 export init=/sbin/init
 export readonly=y
-export rootmnt=/root
+export rootmnt=/root_
 export debug=
 export panic=
 export blacklist=
@@ -232,6 +233,17 @@
 	sleep "$ROOTDELAY"
 fi
 
+DISK=`blkid -o device -l -t LABEL=rootfs | sed 's/p2$//'`
+FIRSTTIME=`dd if=$DISK bs=1 skip=25599 count=1 2>/dev/null`
+if [ $FIRSTTIME = '1' ]; then
+	dd if=/dev/zero of=$DISK bs=1 seek=25599 count=1 conv=notrunc
+	parted -s ${DISK} resizepart 2 100%
+	sleep 1
+	e2fsck -f -p ${DISK}p2
+	resize2fs -f ${DISK}p2
+	tune2fs -m 0.5 ${DISK}p2
+	e2fsck -f -p ${DISK}p2
+fi
 maybe_break premount
 [ "$quiet" != "y" ] && log_begin_msg "Running /scripts/init-premount"
 run_scripts /scripts/init-premount
@@ -251,6 +263,27 @@
 mountroot
 log_end_msg
 
+insmod /usr/lib/modules/*/kernel/fs/squashfs/squashfs.ko
+insmod /usr/lib/modules/*/kernel/fs/overlayfs/overlay.ko
+mount -o remount,rw /root_
+if [ -f /root_/fs/os.new.img ]; then
+	mv /root_/fs/os.new.img /root_/fs/os.img
+fi
+mount -t squashfs -o loop /root_/fs/os.img /root_/fs/lower
+if [ -f /root_/reset.overlay ]; then
+	rm -rf /root_/fs/upper /root_/fs/work /root_/reset.overlay
+	mkdir /root_/fs/upper
+	mkdir /root_/fs/work
+fi
+if [ -f /root_/reset.admin ]; then
+	rm -rf /root_/admin /root_/reset.admin
+fi
+if [ ! -d /root_/admin ]; then
+	cp -a /root_/fs/lower/home/admin.default /root_/admin
+fi
+mount -t overlay -o lowerdir=/root_/fs/lower,upperdir=/root_/fs/upper,workdir=/root_/fs/work,index=on,metacopy=on none /root
+export rootmnt=/root
+
 if read_fstab_entry /usr; then
 	log_begin_msg "Mounting /usr file system"
 	mountfs /usr
